types/chain_id_test.go

```diff
@@ -0,0 +1,646 @@
+package types
+
+import (
+	strings "strings"
+	"testing"
+
+	genutiltypes "github.com/cosmos/cosmos-sdk/x/genutil/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestParseChainIDFromGenesis(t *testing.T) {
+	testCases := []struct {
+		name       string
+		json       string
+		expChainID string
+	}{
+		{
+			"success",
+			`{
+				"state": {
+				  "accounts": {
+					"a": {}
+				  }
+				},
+				"chain_id": "test-chain-id"
+			}`,
+			"test-chain-id",
+		},
+		{
+			"not exist",
+			`{
+				"state": {
+				  "accounts": {
+					"a": {}
+				  }
+				},
+				"chain-id": "test-chain-id"
+			}`,
+			"",
+		},
+		{
+			"invalid type",
+			`{
+				"chain-id":1,
+			}`,
+			"",
+		},
+		{
+			"invalid json",
+			`[ " ': }`,
+			"",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			chain_id, err := ParseChainIDFromGenesis(strings.NewReader(tc.json))
+			if tc.expChainID == "" {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				require.Equal(t, tc.expChainID, chain_id)
+			}
+		})
+	}
+}
+
+func BenchmarkParseChainID(b *testing.B) {
+	// an arbitrary genesis file from a cronos devnet.
+	json := `
```

Comment:
Please commit this JSON to a file under a directory "testdata" so example "testdata/streaming_chain_id.json" and then read it in by file then kindly run b.ResetTimer() even before each sub-benchmark.

Reply:
done